python/cuvs/cuvs/neighbors/hnsw/hnsw.pyx

```diff
@@ -0,0 +1,364 @@
+        Number of CPU threads used to increase search parallelism.
+        When set to 0, the number of threads is automatically determined.
+    """
+
+    cdef cuvsHnswSearchParams params
+
+    def __init__(self, *,
+                 ef=200,
+                 num_threads=0):
+        self.params.ef = ef
+        self.params.num_threads = num_threads
+
+    def __repr__(self):
+        attr_str = [attr + "=" + str(getattr(self, attr))
+                    for attr in [
+                        "ef", "num_threads"]]
+        return "SearchParams(type=HNSW, " + (", ".join(attr_str)) + ")"
+
+    @property
+    def ef(self):
+        return self.params.ef
+
+    @property
+    def num_threads(self):
+        return self.params.num_threads
+
+
+cdef class Index:
+    """
+    HNSW index object. This object stores the trained HNSW index state
+    which can be used to perform nearest neighbors searches.
+    """
+
+    cdef cuvsHnswIndex_t index
+    cdef bool trained
+
+    def __cinit__(self):
+        self.trained = False
+        check_cuvs(cuvsHnswIndexCreate(&self.index))
+
+    def __dealloc__(self):
+        if self.index is not NULL:
+            check_cuvs(cuvsHnswIndexDestroy(self.index))
+
+    @property
+    def trained(self):
+        return self.trained
+
+    def __repr__(self):
+        # todo(dgd): update repr as we expose data through C API
+        attr_str = []
+        return "Index(type=HNSW, metric=L2" + (", ".join(attr_str)) + ")"
+
+
+@auto_sync_resources
+def save(filename, cagra.Index index, resources=None):
+    """
+    Saves the CAGRA index to a file as an hnswlib index.
+
+    Saving / loading the index is experimental. The serialization format is
+    subject to change.
+
+    Parameters
+    ----------
+    filename : string
+        Name of the file.
+    index : Index
+        Trained CAGRA index.
+    {resources_docstring}
+
+    Examples
+    --------
+    >>> import cupy as cp
+    >>> from cuvs.neighbors import cagra
+    >>> n_samples = 50000
+    >>> n_features = 50
+    >>> dataset = cp.random.random_sample((n_samples, n_features),
+    ...                                   dtype=cp.float32)
+    >>> # Build index
+    >>> index = cagra.build(cagra.IndexParams(), dataset)
+    >>> # Serialize and deserialize the cagra index built
+    >>> hnsw.save("my_index.bin", index)
+    """
+    cdef string c_filename = filename.encode('utf-8')
+    cdef cuvsResources_t res = <cuvsResources_t>resources.get_c_obj()
+    check_cuvs(cagra.cuvsCagraSerializeToHnswlib(res,
+                                                 c_filename.c_str(),
+                                                 index.index))
+
+
+@auto_sync_resources
+def load(filename, dim, dtype, metric="sqeuclidean", resources=None):
+    """
+    Loads base-layer-only hnswlib index from file, which was originally
+    saved as a built CAGRA index.
+
+    Saving / loading the index is experimental. The serialization format is
```

Comment:
I know you are alluding to it in this description but we should also be very explicit and mention that the serialized index can only be loaded with the functions in cuVS because the hnswlib version underneath is patched to be able to search the converted CAGRA graph. 


We recently had 2 different partners try to do this conversion themselves and they didn't realixe they couldn't just load this into hnswlib proper.

Reply:
Right above, the note is quite clear:
```
The loaded index is immutable and can only
    be searched by the hnswlib wrapper in cuVS, as the format is not
    compatible with the original hnswlib.
```

> can only be searched by the hnswlib wrapper in cuVS, as the format is not compatible with the original hnswlib.